src/Filecoin/Address.cpp

```diff
@@ -0,0 +1,177 @@
+    if (string.length() > 22)
+        return false;
+    for (int i = 2; i < string.length(); i++) {
+        if (string[i] < '0' || string[i] > '9') {
+            return false;
+        }
+    }
+    try {
+        auto num = std::stoull(string.substr(2));
+        if (num < 0u || num > 18446744073709551615u) {
+            return false;
+        }
+    } catch (...) {
+        return false;
+    }
+    return true;
+}
+
+static bool isValidBase32(const std::string& string, Address::Type type) {
+    // Check if valid Base32.
+    uint8_t size = Address::payload_size(type);
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN)) {
+        return false;
+    }
+
+    // Check size
+    if (decoded.size() != size + 4) {
+        return false;
+    }
+
+    // Extract raw address.
+    Data address;
+    address.push_back(static_cast<uint8_t>(type));
+    address.insert(address.end(), decoded.data(), decoded.data() + size);
+
+    // Verify checksum.
+    Data should_sum = Hash::blake2b(address, checksum_size);
+    return std::memcmp(should_sum.data(), decoded.data() + size, checksum_size) == 0;
+}
+
+bool Address::isValid(const std::string& string) {
+    if (string.length() < 3) {
+        return false;
+    }
+    // Only main net addresses supported.
+    if (string[0] != 'f') {
+        return false;
+    }
+    // Get address type.
+    auto type = parse_type(string[1]);
+    if (!type) {
+        return false;
+    }
+
+    // ID addresses are special, they are just numbers.
+    return type == Type::ID ? isValidID(string) : isValidBase32(string, *type);
+}
+
+Address::Address(const std::string& string) {
+    if (!isValid(string))
+        throw std::invalid_argument("Invalid address data");
+
+    Type type = *(parse_type(string[1]));
+    // First byte is type
+    bytes.push_back(static_cast<uint8_t>(type));
+    if (type == Type::ID) {
+        uint64_t id = std::stoull(string.substr(2));
+        while (id >= 0x80) {
+            bytes.push_back(((uint8_t)id) | 0x80);
+            id >>= 7;
+        }
+        bytes.push_back((uint8_t)id);
+        return;
+    }
+
+    Data decoded;
+    if (!Base32::decode(string.substr(2), decoded, BASE32_ALPHABET_FILECOIN))
+        throw std::invalid_argument("Invalid address data");
+    uint8_t payload_size = Address::payload_size(type);
+
+    bytes.insert(bytes.end(), decoded.data(), decoded.data() + payload_size);
+}
+
+Address::Address(const Data& data) {
+    if (!isValid(data)) {
+        throw std::invalid_argument("Invalid address data");
+    }
+    bytes = data;
+}
+
+Address::Address(const PublicKey& publicKey) {
+    bytes.push_back(static_cast<uint8_t>(Type::SECP256K1));
+    Data hash = Hash::blake2b(publicKey.bytes, payload_size(Type::SECP256K1));
+    bytes.insert(bytes.end(), hash.begin(), hash.end());
+}
+
+std::string Address::string() const {
+    std::string s;
+    // Main net address prefix
+    s.push_back('f');
```

Comment:
I think we can define a constant for `f`

Reply:
I added `Address::PREFIX`